staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/vault/client.go

```diff
@@ -0,0 +1,220 @@
+func newClientWrapper(config *EnvelopeConfig) (*clientWrapper, error) {
+	client, err := newVaultClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Vault transit path is configurable. "path", "/path", "path/" and "/path/"
+	// are the same.
+	transit := "transit"
+	if config.TransitPath != "" {
+		transit = strings.Trim(config.TransitPath, "/")
+	}
+
+	// auth path is configurable. "path", "/path", "path/" and "/path/" are the same.
+	auth := "auth/"
+	if config.AuthPath != "" {
+		auth = strings.Trim(config.AuthPath, "/")
+	}
+	wrapper := &clientWrapper{
+		client:      client,
+		encryptPath: "/v1/" + transit + "/encrypt/",
+		decryptPath: "/v1/" + transit + "/decrypt/",
+		authPath:    auth + "/",
+	}
+
+	// Set token for the api.client.
+	if config.Token != "" {
+		client.SetToken(config.Token)
+	} else {
+		err = wrapper.refreshToken(config, wrapper.version)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapper, nil
+}
+
+func newVaultClient(config *EnvelopeConfig) (*api.Client, error) {
+	vaultConfig := api.DefaultConfig()
+	vaultConfig.Address = config.Address
+
+	tlsConfig := &api.TLSConfig{
+		CACert:        config.CACert,
+		ClientCert:    config.ClientCert,
+		ClientKey:     config.ClientKey,
+		TLSServerName: config.TLSServerName,
+	}
+	if err := vaultConfig.ConfigureTLS(tlsConfig); err != nil {
+		return nil, err
+	}
+
+	return api.NewClient(vaultConfig)
+}
+
+// Get token by login and set the value to api.Client.
+func (c *clientWrapper) refreshToken(config *EnvelopeConfig, version uint) error {
+	c.rwmutex.Lock()
+	defer c.rwmutex.Unlock()
+
+	// The token has been refreshed by other goroutine.
+	if version < c.version {
+		return nil
+	}
+
+	var err error
+	switch {
+	case config.ClientCert != "" && config.ClientKey != "":
+		err = c.tlsToken(config)
+	case config.RoleID != "":
+		err = c.appRoleToken(config)
+	default:
+		err = fmt.Errorf("invalid authentication configuration %+v", config)
+	}
+
+	c.version++
+	return err
+}
+
+func (c *clientWrapper) tlsToken(config *EnvelopeConfig) error {
+	resp, err := c.client.Logical().Write(c.authPath+"cert/login", nil)
+	if err != nil {
+		return err
+	}
+
+	c.client.SetToken(resp.Auth.ClientToken)
+	return nil
+}
+
+func (c *clientWrapper) appRoleToken(config *EnvelopeConfig) error {
+	data := map[string]interface{}{
+		"role_id":   config.RoleID,
+		"secret_id": config.SecretID,
+	}
+	resp, err := c.client.Logical().Write(c.authPath+"approle/login", data)
+	if err != nil {
+		return err
+	}
+
+	c.client.SetToken(resp.Auth.ClientToken)
+	return nil
+}
+
+func (c *clientWrapper) decrypt(keyName string, cipher string) (string, error) {
+	var result string
+
+	data := map[string]string{"ciphertext": cipher}
+	resp, err := c.request(c.decryptPath+keyName, data)
+	if err != nil {
+		return result, err
+	}
+
+	result, ok := resp.Data["plaintext"].(string)
+	if !ok {
+		return result, fmt.Errorf("failed type assertion of vault decrypt response to string")
```

Comment:
Sorry, i meant add the type of the response that was not a string `reflect.TypeOf(resp.Data["plaintext"])`